New Testcaseselection and buffer simulation with now running testcases and seperate functions for Filter and Selection

FabianBer opened the pull request:
Hey @embed-me 
Sorry for opening this new PR but the formatting issue made it necessary to reset by fork and force push to make only new changes visible.

- As discussed in the other PR the buffer functionality now works correctly with your and my tests. The issue was that I did not test it with underscores in the beginning. This should now also word with special characters and does not depend on the file and the entity being named equally. 
- The testcaseselection is now persisten however I had to move this to two seperate functions to account for the addition of new testcases which you might want to select. If you run the last selection but there were no testcases selected the Selection buffer is displayed to prompt for a new selection. For running a new selection i used `n: New Selection`, for running the previously selected testcases I added `l: Last Selection`.
- If no Testcases are selected the function is now terminated and no testcases are being run.
- The filter option has been put back in as you rightly pointed out that is a better way to add new functionality.

Again sorry for the inconvenience with the new pull request but this was the only solution i found for the formatting issue, although my experience with github is limited and ther might have been a better solution.

Best regards,
Fabian

embed-me reviewed (changes requested):
Hi @FabianBer,
Thank you, great work! Only some small remarks from my side.

FabianBer commented on vunit-mode.el:
```diff
@@ -144,6 +153,112 @@
+		   "Exit Testcase Selection without executing Tests")
+    (use-local-map widget-keymap)
+    (widget-setup)
+    (goto-char firstselectionpos)))
```
Yes that also makes the selection change more evident.

FabianBer commented on vunit-mode.el:
```diff
@@ -144,6 +153,112 @@
+	(error "No Testcases were found"))
+    (remove-overlays)
+    ;; insert selection widgets for testcases
+    (widget-insert "Select Testcases below with mouse or Arrow Keys and <Ret>\n\n")
```
You are right general emacs usage works and should be known by the user.

FabianBer commented on vunit-mode.el:
```diff
@@ -112,6 +112,15 @@
 (defconst vunit--regex-testcase "run\s*(\"\\([^\"]*\\)\")"
   "Regular expression for testcases.")
 
+(defvar vunit--testcasestring ""
```
Done.

FabianBer commented on vunit-mode.el:
```diff
@@ -247,13 +386,15 @@
+_f_: List Files        ^ ^              _b_: Buffer           _e_: Keep-Compiling
+_x_: Clean             ^ ^              _t_: Cursor           _p_: Fail-Fast
+^ ^                    ^ ^              _n_: New Selection    _d_: Debug
+^ ^                    ^ ^              _l_: Last Selection   ^ ^
```
Also changed, the initial thought behind last was to better fit with the shortcuts.

FabianBer commented on vunit-mode.el:
```diff
@@ -144,6 +153,112 @@
+  (kill-all-local-variables)
+  ;;get the testcases using vunit
+  (let* (beginpos endpos testcasestring testcases coding-str checkboxes (end nil) firstselectionpos)
+    (insert (shell-command-to-string (format "%s %s %s --output-path %s --no-color --num-threads %d %s"
```
I moved this code to a new function vunit--format-call-string that gets `param` and returns the formatted string.

FabianBer commented on vunit-mode.el:
```diff
@@ -144,6 +153,112 @@
+      (move-end-of-line 1)
+      (setq end (eobp))
+      (move-beginning-of-line 1))
+    (move-beginning-of-line 1)
```
No you are absolutely right the `(move-beginning-of-line 1)` was not necessary.

FabianBer commented on vunit-mode.el:
```diff
@@ -144,6 +153,112 @@
+			     (ignore wid)
+			     (ignore ignore)
+			     (kill-buffer-and-window))
+		   "Exit Testcase Selection without executing Tests")
```
A simple Exit is definetly enough.

FabianBer commented:
In addition to your remarks I also noticed an issue, that the buffer was not being closed correctly if no testcases are found and changed that. To avoid duplication I moved the functionality to a separate function. Additionally I noticed that in this case the previous selection was not being deleted and therefore persisted even if in a new Selection no testcases were found. I changed this also to make sure the testcasestring is reset as soon as list is being called.

embed-me commented:
Hi @FabianBer,
Sorry for my late reply, I was on vacation and did not have access to my computer. I just took a look at your changes, great work, thank you very much for your contribution 👍
